crates/runix/src/flake_ref2/file.rs

```diff
@@ -0,0 +1,225 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::path::Path;
+use std::str::FromStr;
+
+use derive_more::{Deref, Display};
+use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use thiserror::Error;
+use url::Url;
+
+use super::lock::NarHash;
+use super::protocol::{self, Protocol, WrappedUrl, WrappedUrlParseError};
+use super::{BoolReprs, FlakeRefSource};
+
+pub type FileUrl<Protocol> = WrappedUrl<Protocol>;
+
+/// <https://cs.github.com/NixOS/nix/blob/f225f4307662fe9a57543d0c86c28aa9fddaf0d2/src/libfetchers/tarball.cc#L287>
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq)]
+#[serde(tag = "file")]
+pub struct FileRef<Protocol: FileProtocol, A: ApplicationProtocol = File> {
+    url: WrappedUrl<Protocol>,
+
+    #[serde(skip)]
+    _type: Application<A>,
+
+    #[serde(flatten)]
+    attributes: FileAttributes,
+}
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq)]
+pub struct FileAttributes {
+    #[serde(flatten)]
+    nar_hash: Option<NarHash>,
+
+    unpack: Option<Unpack>,
+
+    name: Option<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Deref)]
+struct Unpack(#[serde(deserialize_with = "BoolReprs::deserialize_bool")] bool);
+
+pub trait ApplicationProtocol: Default {
+    fn protocol() -> Cow<'static, str>;
+    fn required(url: &Url) -> bool;
+}
+
+#[derive(Debug, Default, PartialEq, Eq, DeserializeFromStr, SerializeDisplay)]
+pub struct Application<P: ApplicationProtocol> {
+    inner: P,
+}
+
+#[derive(Debug, Error)]
+#[error("Invalid Application '{0}', expected {1}")]
+pub struct InvalidApplication<T: ApplicationProtocol>(String, Application<T>);
+
+impl<T: ApplicationProtocol> FromStr for Application<T> {
+    type Err = InvalidApplication<T>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s != T::protocol() {
+            return Err(InvalidApplication(s.to_string(), Self::default()));
+        }
+        Ok(Self::default())
+    }
+}
+
+impl<T: ApplicationProtocol> Display for Application<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", T::protocol())
+    }
+}
+
+#[derive(Display, Default, Debug)]
+pub struct File;
+impl ApplicationProtocol for File {
+    fn protocol() -> Cow<'static, str> {
+        "file".into()
+    }
+
+    fn required(url: &Url) -> bool {
+        let is_tarball_url = Path::new(url.path())
+            .file_name()
+            .map(|name| {
+                [
+                    ".zip", ".tar", ".tgz", ".tar.gz", ".tar.xz", ".tar.bz2", ".tar.zst",
+                ]
+                .iter()
+                .any(|ext| name.to_string_lossy().ends_with(ext))
+            })
+            .unwrap_or_default();
```

Comment:
This could be a function also used in Tarball

Reply:
Yeah i guess we can reuse that